### PR TITLE
74 - Add diff renderer and improve reporting

### LINKED ISSUE #74: Diff generator
This function should generate a diff, optionally with ANSI colour codes via a boolean parameter, between two values. It should support deeply nested tables, accounting for circular references.

## Acceptance criteria

- It takes two arguments: `valueBefore` and `valueAfter`
- It prints a table with correct indentation
- For fields that are equal between the two tables, it prints the field only once and with no colouring / annotation
- For fields that are different between the two tables
  - It prints the field in the before value
    - highlighted in red
    - with a `-` before it - this should be at the same horizontal position regardless of indentation
  - It _also_ prints the field in the after value
    - highlighted in green
    - with a `+` before it - this should follow the same positioning as above
  - It prints the before and after values adjacent to each other vertically
  - It prints the before value before the after value
- It is used in matchers that can compare two table values, such as `toEqual`

### PR DESCRIPTION
**Pull request by Derpius**
## Issue

Resolves #74

## Changes

- Repurposed `prettyValue` into `serialiseValue`, which recursively prints tables that do not have an overridden `tostring`
  - This is aligned with Jest
- Added a new `renderDiff` function which is used to generate a diff between any two values, optionally performing a "deep" recursive diff
- Updated all code using `prettyValue` to use `serialiseValue`
- Changed `require` paths to remove the need for `src.lua.` in every `require()`
  - Paths should be changed in any files that would be included in the diff regardless
- Fixed matcher assertions printing the stack trace twice
- Changed the test result error to not colour its message body
  - This allows matchers to colour their own message (needed for `renderDiff`)
- Changed the `expect` print to not serialise the values into `expect(...).matcherName(...)` to avoid duplicating information

## Impact

- Significantly improves the user experience of Lest by allowing users to immediately see deep inequalities in tables through the diff
- Improves the reporting for non-deep equality tests to better highlight expected and received values
- All existing test failure messages will change colour from red to white
  - #114 to add colours back in
- DevX improved by removing the need to import from `src.lua` manually, and automatically importing `index.lua` files when a directory path is given
  - #115 to update remaining imports

## Testing

Fairly well covered by automated tests, but should try the following:

- `toBe` failure messages display serialised expected and received with no table diff rendering
- `toEqual` failure messages match `toBe` if one or both values are not a table
- `toEqual` failure messages for two tables
  - handles cyclic references
  - shows correct expected and received counts for the number of differences
  - correctly chooses whether a value can be added to the table as an array item (`{1, 2, 3}`), symbol key (`{foo = "bar"}`) or expression key (`{["end"] = "bar"}`)
  - correctly marks fields as either unchanged, added, removed or changed (added and removed adjacent for same key)
- `never.toBe` and `never.toEqual` failure messages always show the serialised expected value in the format `Expected: not %s`

## Helpful Links

[API Docs](https://taservers.github.io/lest/docs/api/expect)

[Project Board](https://github.com/orgs/TAServers/projects/1)

[Discord](https://discord.tasevers.com)

### CODE REVIEW

**Comment by yogwoggf** [comment]
If @Vurv78 doesn't pick this up later today then I'll probably start a review

File: packages/lest/src/lua/utils/is-lua-symbol.lua [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	["while"] = true,
+}
+
+--- Returns true if the value is a string containing a valid Lua symbol
+---@param value any
+---@return boolean
+local function isLuaSymbol(value)
+	return type(value) == "string"
+		and not KEYWORDS[value]
+		and not not string.match(value, "^[_%a][_%a%d]*$")
```
Comment (yogwoggf) [inline packages/lest/src/lua/utils/is-lua-symbol.lua:31]: Just to confirm, this is a boolean coercion right? Looks sort of strange

**Review by yogwoggf — APPROVED** [review]
LGTM 👍

File: packages/lest/src/lua/utils/is-lua-symbol.lua (hunk 1 above)
Comment (Derpius) [reply packages/lest/src/lua/utils/is-lua-symbol.lua:31]: Yep, to avoid returning a non-boolean here

**Pull request merged**